llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.h

```diff
@@ -69,7 +69,53 @@ class SPIRVGlobalRegistry {
                              AQ::AccessQualifier accessQual = AQ::ReadWrite,
                              bool EmitIR = true);
 
+  const MachineFunction *MetaMF;
+
+  // Maps a local register to the corresponding global aliases in meta-function.
+  using LocalToGlobalRegTable = std::map<Register, Register>;
+  using RegisterAliasMapTy =
+      std::map<const MachineFunction *, LocalToGlobalRegTable>;
+
+  // The table contains global aliases of local registers for each machine
+  // function. They are collected when local instructions are moved to global
+  // level on SPIRVGlobalTypesAndRegNumPass. Then the aliases are used
+  // to substitute global aliases during code emission.
+  RegisterAliasMapTy RegisterAliasTable;
+
+  // The set contains machine instructions which are necessary for correct MIR
+  // but will not be finally emitted.
+  DenseSet<MachineInstr *> InstrsToDelete;
```

Comment:
I don't like the idea of keeping this list through to the `AsmPrinter`. But currently I don't see any other way to keep this info as well. We could use some metadata flag, but on MIR level that would require setupping new `MIRBuilder` for each instruction (at least that how it looks at the first glimpse). We also have `MachineInstr::AsmPrinterFlags`, but the comment for it suggests not to keep any semantic info in it.

Maybe the community would suggest something?

Reply:
I agree with you, it's better to make a standard technique for this but I have not found it too. I suggest to publish this in review and receive feedback from the community.